include/ada/parser.h

```diff
@@ -21,9 +22,18 @@ namespace ada::parser {
   /**
    * Parses a url.
    */
-  url parse_url(std::string_view user_input,
-                const ada::url* base_url = nullptr,
-                ada::encoding_type encoding = ada::encoding_type::UTF8);
+  template <class result_type = url>
+  result_type parse_url(std::string_view user_input,
+                        const ada::url* base_url = nullptr,
```

Comment:
for the base_url we probably want to define a type that provides what we need without requiring that it an ada::url.